rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,214 @@
+///     type Item = u32;
+///     type Iterator = alloc::vec::IntoIter<u32>;
+///
+///     fn start(arg: &self) -> Option<Box<Peekable<Self::Iterator>>> {
+///         Box::try_new(arg.contents.clone().into_iter().peekable()).ok()
+///     }
+/// }
+/// ```
+///
+/// [`seq_operations`]: ../../../include/linux/seq_file.h
+pub trait SeqOperations {
+    /// Type produced on each iteration.
+    type Item: Display;
+
+    /// Type created when the seq file is opened.
+    type Iterator: Iterator<Item = Self::Item>;
+
+    /// Called once each time the `seq_file` is opened.
+    fn start(arg: &Self) -> Option<Box<Peekable<Self::Iterator>>>;
+}
+
+extern "C" fn stop_callback<T: SeqOperations>(
+    _m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) {
+    if !v.is_null() {
+        // SAFETY: `v` was created by a previous call to `next_callback` or
+        // `start_callback` and both functions return either a null pointer
+        // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+        let _iterator = unsafe { Box::from_raw(v as *mut Peekable<T::Iterator>) };
+    }
+}
+
+extern "C" fn next_callback<T: SeqOperations>(
+    _m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    if !v.is_null() {
+        // SAFETY: `v` was created by a previous call to `next_callback` or
+        // `start_callback` and both functions return either a null pointer
+        // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+        //
+        // The caller guarantees tha `pos` is a valid pointer to an
+        // `loff_t` and expects this function to mutate the value.
+        let mut iterator = unsafe {
+            *pos += 1;
+            Box::from_raw(v as *mut Peekable<T::Iterator>)
+        };
+        match iterator.as_mut().next() {
+            Some(_seen) => Box::into_raw(iterator) as *mut c_types::c_void,
+            None => ptr::null_mut(),
+        }
+    } else {
+        ptr::null_mut()
+    }
+}
+
+extern "C" fn show_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) -> c_types::c_int {
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+    let iterator = unsafe { (v as *mut Peekable<T::Iterator>).as_mut() };
+    if let Some(iterator) = iterator {
+        let s = match iterator.peek() {
+            // TODO: Replace with fallible `to_string` when available.
+            Some(item) => item.to_string() + "\0",
+            None => "\0".to_string(),
+        };
+        // SAFETY: Calling a C function. `s` is guaranteed to be null terminated
+        // because we explicitly constructed it just above.
+        unsafe {
+            bindings::seq_puts(m, s.as_ptr() as *const u8 as *const c_types::c_char);
```

Comment:
Can we use `seq_printf` here and specify the string length so that we don't need to depend on `to_string` and concatenation?

Reply:
Looked into it a bit, but I can't really see how to do it (at least not in a way that's generic over `Display`). Here are some options I found:
1. If we use `seq_printf` with the length in the format string, we'll still need to allocate a `String` to calculate the length and get the pointer to pass. It would get rid of the concat though.
1. Instead of having `Item: Display`, we could require a `fn(&Item) -> &str`. Then the implementer can decide if they want to allocate a `String` or not.
1. We could have the trait define a print function that take `Item` and `seq_file` and let the implementer use the various `seq` print functions. But then they'll need to use `unsafe` and they'd have the same problem as us if they want to print a string or type that doesn't have a `printf` format.
1. Instead of having `Item: Display` we could make our own trait for types that can be formatted with `printf` and write safe wrappers around `seq_printf` for each of those.
1. There is the [VaList](https://doc.rust-lang.org/std/ffi/struct.VaList.html) type which seems to handle varargs, so we could have that type instead of the generic `Item`, but it seems more for defining Rust code that can be called on the C side with varargs, not for generating a Rust type that can be passed to C functions expecting varargs.

I don't think there's any support for putting a dynamically sized string on the stack.

Reply:
Went with 2.

Reply:
I think it's fine for now. In the future if there's a need for it, I think it would make sense to do a variant of option 3, where we would provide abstractions that would allow users to use the various `seq_file` write variants without having to call unsafe functions.

Reply:
Happy to make a follow-up PR for that.